torch_geometric/graphgym/logger.py

```diff
@@ -247,3 +258,105 @@ def create_logger():
+        lr = trainer.lr_scheduler_configs[0].scheduler.get_last_lr()[0]
+        stats = dict(true=true.detach().cpu(), pred=pred.detach().cpu(),
+                     pred_score=pred_score.detach().cpu(), loss=loss.item(),
+                     lr=lr, time_used=time.time() - time_start,
+                     params=cfg.params)
+
+        return stats
+
+    def on_train_epoch_start(self, trainer: "pl.Trainer",
+                             pl_module: "pl.LightningModule") -> None:
+        self._train_epoch_start_time = time.time()
+
+    def on_validation_epoch_start(self, trainer: "pl.Trainer",
+                                  pl_module: "pl.LightningModule") -> None:
+        self._val_epoch_start_time = time.time()
+
+    def on_test_epoch_start(self, trainer: "pl.Trainer",
+                            pl_module: "pl.LightningModule") -> None:
+        self._test_epoch_start_time = time.time()
+
+    def on_train_batch_end(
+        self,
+        trainer: "pl.Trainer",
+        pl_module: "pl.LightningModule",
+        outputs: Dict[str, Any],
+        batch: Any,
+        batch_idx: int,
+        unused: int = 0,
+    ) -> None:
+        stats = self._create_stats(self._train_epoch_start_time, outputs,
+                                   trainer)
+        self.train_logger.update_stats(**stats)
+
+    def on_validation_batch_end(
+        self,
+        trainer: "pl.Trainer",
+        pl_module: "pl.LightningModule",
+        outputs: Optional[Dict[str, Any]],
+        batch: Any,
+        batch_idx: int,
+        dataloader_idx: int,
+    ) -> None:
+        stats = self._create_stats(self._val_epoch_start_time, outputs,
+                                   trainer)
+        self.val_logger.update_stats(**stats)
+
+    def on_test_batch_end(
+        self,
+        trainer: "pl.Trainer",
+        pl_module: "pl.LightningModule",
+        outputs: Optional[Dict[str, Any]],
+        batch: Any,
+        batch_idx: int,
+        dataloader_idx: int,
+    ) -> None:
+        stats = self._create_stats(self._test_epoch_start_time, outputs,
+                                   trainer)
+        self.test_logger.update_stats(**stats)
+
+    def on_train_epoch_end(self, trainer: "pl.Trainer",
+                           pl_module: "pl.LightningModule") -> None:
+        self.train_logger.write_epoch(trainer.current_epoch)
+
+    def on_validation_epoch_end(self, trainer: "pl.Trainer",
+                                pl_module: "pl.LightningModule") -> None:
+        self.val_logger.write_epoch(trainer.current_epoch)
+
+    def on_test_epoch_end(self, trainer: "pl.Trainer",
+                          pl_module: "pl.LightningModule") -> None:
+        self.test_logger.write_epoch(trainer.current_epoch)
+
+    def on_epoch_end(self, trainer: "pl.Trainer",
+                     pl_module: "pl.LightningModule") -> None:
+        for logger in self._logger:
+            logger.close()
+
+    @property
+    def train_logger(self):
```

Comment:
Just wondering if "create_loggers" should return a dict, this seems a tiny bit error prone.

Reply:
Agree! Since this PR doesn't update the training epoch implementation so I would keep this change for the next PR with the `Trainer` integration. Thoughts @rusty1s @Borda?